### PR TITLE
feature(core): add onPaste and onDrop events to editor

### PR DESCRIPTION
**Pull request by bdbch**
## Please describe your changes

[This message](https://discord.com/channels/818568566479257641/818569721934774272/1202162639317508158) on our Discord server made me realize that there's apparently no easy way to just add a simple paste or drag event handler to the actual editor instance. This user tried to add it to the editors DOM element which works sometimes (doesn't work properly if there's no content inside the contenteditable yet).

## How did you accomplish your changes

I created two very simple plugins to handle paste and drop events passing up their relative events to the option passed through.

Those can be configured on the editor constructor.

## How have you tested your changes

I tried to add `onPaste` and `onDrop` on a local demo.

## How can we verify your changes

See above

## Remarks

No remarks from my side

## Checklist

- [x] The changes are not breaking the editor
- [ ] Added tests where possible
- [x] Followed the guidelines
- [x] Fixed linting issues

## Related issues

No related issues

### CODE REVIEW

File: packages/core/src/plugins/DropPlugin.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import { Plugin, PluginKey } from '@tiptap/pm/state'
+
+export const DropPlugin = (onPaste: (e: DragEvent) => void) => {
+  return new Plugin({
+    key: new PluginKey('tiptap__drop'),
```
Comment (svenadlung) [inline packages/core/src/plugins/DropPlugin.ts:5]: Maybe keep this in lowerCamelCase (`tiptapDrop`), like we do in other plugins?

File: packages/core/src/plugins/PastePlugin.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import { Plugin, PluginKey } from '@tiptap/pm/state'
+
+export const PastePlugin = (onPaste: (e: ClipboardEvent) => void) => {
+  return new Plugin({
+    key: new PluginKey('tiptap__paste'),
```
Comment (svenadlung) [inline packages/core/src/plugins/PastePlugin.ts:5]: Same here.

File: packages/core/src/plugins/DropPlugin.ts (hunk 1 above)
Comment (svenadlung) [reply packages/core/src/plugins/DropPlugin.ts:5]: Or maybe call it `dropEvent`, analog to `focusEvents`?

File: packages/core/src/index.ts [hunk 3]
```diff
@@ -11,6 +11,8 @@ export * from './NodePos.js'
 export * from './NodeView.js'
 export * from './PasteRule.js'
 export * from './pasteRules/index.js'
+export * from './plugins/DropPlugin.js'
```
[the remark this answers is not in the document]
Comment (svenadlung) [reply packages/core/src/index.ts:14]: Oh, I just realized focusEvents is a wrapped plugin (as a Tiptap extension). But maybe we keep it the same style?

**Review by svenadlung — COMMENTED** [review]
Please check my comments.

File: packages/core/src/plugins/DropPlugin.ts (hunk 1 above)
Comment (bdbch) [reply packages/core/src/plugins/DropPlugin.ts:5]: how about `tiptap__dropEvent` - I'm using the `tiptap__` prefix to make sure to never accidentially conflict with other pluginKeys

File: packages/core/src/plugins/PastePlugin.ts (hunk 2 above)
Comment (bdbch) [reply packages/core/src/plugins/PastePlugin.ts:5]: how about `tiptap__pasteEvent` - I'm using the `tiptap__` prefix to make sure to never accidentially conflict with other pluginKeys

File: packages/core/src/plugins/DropPlugin.ts (hunk 1 above)
Comment (nperez0111) [reply packages/core/src/plugins/DropPlugin.ts:5]: PluginKeys are symbols by default no? So this should only ever exist once.

File: packages/core/src/plugins/DropPlugin.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+    props: {
+      handleDrop: (_, e) => {
+        onPaste(e)
```
Comment (nperez0111) [inline packages/core/src/plugins/DropPlugin.ts:9]:
```suggestion
        onDrop(e)
```

**Review by nperez0111 — APPROVED** [review]

File: packages/core/src/types.ts [hunk 5]
```diff
@@ -121,6 +121,8 @@ export interface EditorOptions {
   onFocus: (props: EditorEvents['focus']) => void;
   onBlur: (props: EditorEvents['blur']) => void;
   onDestroy: (props: EditorEvents['destroy']) => void;
+  onPaste?: (e: ClipboardEvent) => void
+  onDrop?: (e: DragEvent) => void
```
Comment (nperez0111) [inline packages/core/src/types.ts:125]: I think these types are intentionally not optional, there is a Partial<EditorOptions> wherever we use it

File: packages/core/src/plugins/PastePlugin.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import { Plugin, PluginKey } from '@tiptap/pm/state'
+
+export const PastePlugin = (onPaste: (e: ClipboardEvent) => void) => {
+  return new Plugin({
+    key: new PluginKey('tiptapPaste'),
+
+    props: {
+      handlePaste: (_, e) => {
+        onPaste(e)
+      },
```
Comment (nperez0111) [inline packages/core/src/plugins/PastePlugin.ts:10]: On looking at this again, would we want to also pass through any of these other values?

I see that `handlePaste` is defined as:
```
    /**
    Can be used to override the behavior of pasting. `slice` is the
    pasted content parsed by the editor, but you can directly access
    the event to get at the raw content.
    */
    handlePaste?: (this: P, view: EditorView, event: ClipboardEvent, slice: Slice) => boolean | void;
```
I think at least `slice` might be useful to passthrough?

File: packages/core/src/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (nperez0111) [reply packages/core/src/index.ts:14]: I think that is fair, I'm okay with this organization

File: packages/core/src/plugins/PastePlugin.ts (hunk 6 above)
Comment (bdbch) [reply packages/core/src/plugins/PastePlugin.ts]: Good idea - done.
Added `slice` for onPaste and `slice` and `moved` for onDrop.

**Comment by YousefED** [comment]
It seems a bit weird that these methods are initialized to `() => null` and a bit later in the code they're checked with `if(this.options.onPaste) {`.

imo it would make more sense to initialize them as undefined

**Comment by nperez0111** [comment]
> It seems a bit weird that these methods are initialized to `() => null` and a bit later in the code they're checked with `if(this.options.onPaste) {`.
> 
> imo it would make more sense to initialize them as undefined

Good point, resolved with: https://github.com/ueberdosis/tiptap/pull/5684

**Pull request merged**